File: rlm/cli/context.py
from __future__ import annotations

import os
import shutil
import sys
from dataclasses import dataclass, field
from pathlib import Path
from typing import MutableMapping
from urllib.parse import urlparse

# Backward-compat: re-export from new homes so existing "from rlm.cli.context import X" still works
from rlm.cli.output import print_error, print_success  # noqa: F401
from rlm.cli.checks import doctor_runtime_requirement, require_supported_runtime  # noqa: F401


def _discover_project_root(start: Path) -> Path:
    current = start.resolve()
    for candidate in (current, *current.parents):
        if (candidate / "pyproject.toml").exists() or (candidate / ".git").exists():
            return candidate
    return current


def resolve_operator_api_base_url(env: MutableMapping[str, str], api_base_url: str) -> str:
    operator_host = str(env.get("RLM_OPERATOR_HOST", "") or "").strip()
    if operator_host:
        return operator_host.rstrip("/")

    configured = str(api_base_url or "").strip() or "http://127.0.0.1:5000"
    parsed = urlparse(configured if "://" in configured else f"http://{configured}")
    scheme = parsed.scheme or "http"
    host = str(parsed.hostname or env.get("RLM_API_HOST", "127.0.0.1") or "127.0.0.1").strip() or "127.0.0.1"
    if host == "0.0.0.0":
        host = "127.0.0.1"

    try:
        port = parsed.port or 5000
    except ValueError:
        port = 5000

    if port <= 0 or port > 65535:
        port = 5000

    return f"{scheme}://{host}:{port}"


@dataclass(frozen=True, slots=True)
class CliPaths:
    cwd: Path
    home: Path
    project_root: Path
    state_root: Path
    launcher_state_path: Path
    runtime_dir: Path
    log_dir: Path
    env_path: Path
    skills_dir: Path

    @classmethod
    def discover(cls, *, cwd: Path | None = None, home: Path | None = None, env: MutableMapping[str, str] | None = None) -> CliPaths:
        current_cwd = (cwd or Path.cwd()).resolve()
        current_home = (home or Path.home()).resolve()
        project_root = _discover_project_root(current_cwd)
        state_root = current_home / ".rlm"
        env_path = current_cwd / ".env"
        if not env_path.exists():
            env_path = state_root / ".env"
        current_env = env or os.environ
        configured_skills_dir = current_env.get("RLM_SKILLS_DIR", "").strip()
        skills_dir = Path(configured_skills_dir).resolve() if configured_skills_dir else project_root / "rlm" / "skills"
        return cls(
            cwd=current_cwd,
            home=current_home,
            project_root=project_root,
            state_root=state_root,
            launcher_state_path=state_root / "launcher-state.json",
            runtime_dir=state_root / "run",
            log_dir=state_root / "logs",
            env_path=env_path,
            skills_dir=skills_dir,
        )


@dataclass(slots=True)
class CliContext:
    env: MutableMapping[str, str] = field(default_factory=lambda: dict(os.environ))
    cwd: Path = field(default_factory=Path.cwd)
    home: Path = field(default_factory=Path.home)
    paths: CliPaths = field(init=False)

    def __post_init__(self) -> None:
        self.paths = CliPaths.discover(cwd=self.cwd, home=self.home, env=self.env)

    @classmethod
    def from_environment(cls, *, load_env: bool = True) -> CliContext:
        context = cls()
        if load_env:
            context.load_env_file(override=False)
        return context

    def resolve_env_path(self) -> Path:
        return self.paths.env_path

    def refresh_paths(self) -> CliPaths:
        self.paths = CliPaths.discover(cwd=self.cwd, home=self.home, env=self.env)
        return self.paths

    def load_env_file(self, *, override: bool = False) -> Path | None:
        env_path = self.resolve_env_path()
        if not env_path.exists():
            return None

        try:
            from dotenv import load_dotenv

            load_dotenv(str(env_path), override=override)
        except ImportError:
            for raw_line in env_path.read_text(encoding="utf-8").splitlines():
                line = raw_line.strip()
                if not line or line.startswith("#") or "=" not in line:
                    continue
                key, value = line.split("=", 1)
                key = key.strip()
                value = value.strip()
                if override or key not in self.env:
                    self.env[key] = value

        self.refresh_paths()
        return env_path

    def has_tool(self, name: str) -> bool:
        return shutil.which(name) is not None

    def api_host(self) -> str:
        return self.env.get("RLM_API_HOST", "127.0.0.1")

    def api_port(self) -> int:
        raw = self.env.get("RLM_API_PORT", "5000")
        try:
            port = int(raw)
        except (ValueError, TypeError):
            port = 5000
        if port < 1 or port > 65535:
            port = 5000
        return port

    def ws_host(self) -> str:
        return self.env.get("RLM_WS_HOST", self.api_host())

    def ws_port(self) -> int:
        return int(self.env.get("RLM_WS_PORT", "8765"))

    def api_base_url(self) -> str:
        return f"http://{self.api_host()}:{self.api_port()}"

    def operator_api_base_url(self) -> str:
        return resolve_operator_api_base_url(self.env, self.api_base_url())

    def ws_base_url(self) -> str:
        return f"ws://{self.ws_host()}:{self.ws_port()}"

    def docs_url(self) -> str:
        return f"{self.api_base_url()}/docs"

    def webchat_url(self) -> str:
        return f"{self.api_base_url()}/webchat"

File: packages/cli/src/context.ts
/**
 * context.ts — Caminhos e contexto do CLI Arkhe.
 *
 * Porta fiel de rlm/cli/context.py
 */

import {
  existsSync,
  readFileSync,
  readdirSync,
} from "node:fs";
import { join, resolve, dirname } from "node:path";
import { homedir } from "node:os";

// ---------------------------------------------------------------------------
// CliPaths
// ---------------------------------------------------------------------------

export interface CliPaths {
  cwd: string;
  home: string;
  projectRoot: string;
  stateRoot: string;
  launcherStatePath: string;
  runtimeDir: string;
  logDir: string;
  envPath: string;
  skillsDir: string;
}

function discoverProjectRoot(start: string): string {
  let current = resolve(start);
  // Percorre até a raiz procurando pyproject.toml ou .git
  while (true) {
    if (
      existsSync(join(current, "pyproject.toml")) ||
      existsSync(join(current, "package.json")) ||
      existsSync(join(current, ".git"))
    ) {
      return current;
    }
    const parent = dirname(current);
    if (parent === current) break;
    current = parent;
  }
  return resolve(start);
}

export function discoverCliPaths(opts: {
  cwd?: string;
  home?: string;
  env?: NodeJS.ProcessEnv;
}): CliPaths {
  const cwd = resolve(opts.cwd ?? process.cwd());
  const home = opts.home ?? homedir();
  const env = opts.env ?? process.env;
  const projectRoot = discoverProjectRoot(cwd);
  const stateRoot = join(home, ".rlm");

  // Resolve .env: primeiro tenta cwd/.env, depois stateRoot/.env
  let envPath = join(cwd, ".env");
  if (!existsSync(envPath)) {
    envPath = join(stateRoot, ".env");
  }

  const configuredSkillsDir = (env["RLM_SKILLS_DIR"] ?? "").trim();
  const skillsDir = configuredSkillsDir
    ? resolve(configuredSkillsDir)
    : join(projectRoot, "rlm", "skills");

  return {
    cwd,
    home,
    projectRoot,
    stateRoot,
    launcherStatePath: join(stateRoot, "launcher-state.json"),
    runtimeDir: join(stateRoot, "run"),
    logDir: join(stateRoot, "logs"),
    envPath,
    skillsDir,
  };
}

// ---------------------------------------------------------------------------
// CliContext
// ---------------------------------------------------------------------------

export class CliContext {
  env: NodeJS.ProcessEnv;
  cwd: string;
  home: string;
  paths: CliPaths;

  constructor(opts: { cwd?: string; home?: string; env?: NodeJS.ProcessEnv } = {}) {
    this.cwd = resolve(opts.cwd ?? process.cwd());
    this.home = opts.home ?? homedir();
    this.env = { ...process.env, ...(opts.env ?? {}) };
    this.paths = discoverCliPaths({ cwd: this.cwd, home: this.home, env: this.env });
  }

  static fromEnvironment(opts: { loadEnv?: boolean } = {}): CliContext {
    const ctx = new CliContext();
    if (opts.loadEnv !== false) {
      ctx.loadEnvFile({ override: false });
    }
    return ctx;
  }

  refreshPaths(): CliPaths {
    this.paths = discoverCliPaths({ cwd: this.cwd, home: this.home, env: this.env });
    return this.paths;
  }

  /** Carrega variáveis de um .env sem depender de dotenv (zero-dep fallback). */
  loadEnvFile(opts: { override?: boolean } = {}): string | null {
    const envPath = this.paths.envPath;
    if (!existsSync(envPath)) return null;

    const lines = readFileSync(envPath, "utf8").split("\n");
    for (const rawLine of lines) {
      const line = rawLine.trim();
      if (!line || line.startsWith("#") || !line.includes("=")) continue;
      const eqIdx = line.indexOf("=");
      const key = line.slice(0, eqIdx).trim();
      let value = line.slice(eqIdx + 1).trim();
      // Remove aspas envolventes
      if ((value.startsWith('"') && value.endsWith('"')) ||
          (value.startsWith("'") && value.endsWith("'"))) {
        value = value.slice(1, -1);
      }
      if (opts.override || !(key in this.env)) {
        this.env[key] = value;
        process.env[key] = value;
      }
    }

    this.refreshPaths();
    return envPath;
  }

  hasTool(name: string): boolean {
    // Detecta se o binário está no PATH
    const { execSync } = require("node:child_process") as typeof import("node:child_process");
    const cmd = process.platform === "win32" ? `where ${name}` : `which ${name}`;
    try {
      execSync(cmd, { stdio: "pipe" });
      return true;
    } catch {
      return false;
    }
  }

  apiHost(): string {
    return this.env["RLM_API_HOST"] ?? "127.0.0.1";
  }

  apiPort(): number {
    return parseInt(this.env["RLM_API_PORT"] ?? "5000", 10);
  }

  wsHost(): string {
    return this.env["RLM_WS_HOST"] ?? this.apiHost();
  }

  wsPort(): number {
    return parseInt(this.env["RLM_WS_PORT"] ?? "8765", 10);
  }

  apiBaseUrl(): string {
    return `http://${this.apiHost()}:${this.apiPort()}`;
  }

  wsBaseUrl(): string {
    return `ws://${this.wsHost()}:${this.wsPort()}`;
  }

  docsUrl(): string {
    return `${this.apiBaseUrl()}/docs`;
  }

  webchatUrl(): string {
    return `${this.apiBaseUrl()}/webchat`;
  }
}
